File: agent_foundation/state.py
"""Session state for the Agent foundation prototype."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .contracts import (
    AnalysisResult,
    FaceConsentState,
    FaceMode,
    RecommendationResponse,
    Route,
    SceneType,
    SessionStatus,
    UserNeedConstraints,
)


@dataclass(slots=True)
class AgentState:
    session_id: str
    scene_type: SceneType = SceneType.MIRROR
    store_id: str = "store_001"
    status: SessionStatus = SessionStatus.COMMUNICATING
    route: Route = Route.UNCLEAR
    analysis: AnalysisResult = field(
        default_factory=lambda: AnalysisResult(
            matched_body_template_id="body_template_03",
            current_style=["casual"],
            confidence=0.82,
        )
    )
    consent: FaceConsentState = field(
        default_factory=lambda: FaceConsentState(
            consent_given=False,
            face_mode=FaceMode.DEFAULT_FACE,
        )
    )
    user_need: UserNeedConstraints = field(default_factory=UserNeedConstraints)
    recommendation_round: int = 0
    shown_set_ids: list[str] = field(default_factory=list)
    selected_set_id: str | None = None
    feedback_history: list[dict[str, Any]] = field(default_factory=list)
    loop_status: str = "collecting"
    last_recommendation: RecommendationResponse | None = None
    last_tool_result: dict[str, Any] | None = None
    errors: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "scene_type": self.scene_type.value,
            "store_id": self.store_id,
            "status": self.status.value,
            "route": self.route.value,
            "analysis": self.analysis.to_dict(),
            "consent": self.consent.to_dict(),
            "user_need": self.user_need.to_dict(),
            "recommendation_round": self.recommendation_round,
            "shown_set_ids": list(self.shown_set_ids),
            "selected_set_id": self.selected_set_id,
            "feedback_history": list(self.feedback_history),
            "loop_status": self.loop_status,
            "last_tool_result": self.last_tool_result,
            "errors": list(self.errors),
        }

File: server/agent/state.ts
import type {
  AgentConstraints,
  AgentLanguage,
  AhaDemoState,
  AnalysisHandoff,
  ConstraintDelta,
  FaceConsentState,
  ParsedNeed,
  RecommendationSet,
  Route,
  SceneType,
  SessionStatus,
  ToolCallRecord
} from "./contracts.js";

export interface FeedbackHistoryEntry {
  set_id: string;
  feedback: Record<string, unknown>;
  constraint_delta: ConstraintDelta;
  created_at: string;
}

export type ReservationState =
  | { reservation_id: string; status: string }
  | { ok: false; reason: string; shortfalls: Array<{ product_id: string; size: string; requested: number; available: number }> };

export interface StoreRouteState {
  summary: string;
  stops: Array<{ product_id: string; name: string; area: string; shelf: string }>;
}

export interface AgentState {
  session_id: string;
  scene_type: SceneType;
  store_id: string;
  // Customer-facing output language. Optional because sessions persisted before
  // this field existed restore without it; readers must fall back to "zh".
  language?: AgentLanguage;
  status: SessionStatus;
  route: Route;
  analysis?: AnalysisHandoff;
  matched_body_template_id?: string;
  user_need?: ParsedNeed;
  constraints: AgentConstraints;
  recommendation_round: number;
  shown_set_ids: string[];
  recommendation_sets: RecommendationSet[];
  selected_set_id?: string;
  feedback_history: FeedbackHistoryEntry[];
  camera_processing_consent: boolean;
  reservation?: ReservationState;
  store_route?: StoreRouteState;
  face: FaceConsentState;
  aha_demo: AhaDemoState;
  lucy_session_status: "idle" | "ready" | "previewing" | "stopped" | "failed";
  loop_status: "idle" | "clarifying" | "active" | "confirmed" | "staff_takeover";
  tool_calls: ToolCallRecord[];
  errors: string[];
}

export function createInitialAgentState(input: {
  session_id: string;
  scene_type: SceneType;
  store_id: string;
  language?: AgentLanguage;
  analysis?: AnalysisHandoff;
}): AgentState {
  return {
    session_id: input.session_id,
    scene_type: input.scene_type,
    store_id: input.store_id,
    language: input.language ?? "zh",
    status: "communicating",
    route: "unclear",
    analysis: input.analysis,
    constraints: {
      prefer: [],
      avoid: []
    },
    recommendation_round: 0,
    shown_set_ids: [],
    recommendation_sets: [],
    feedback_history: [],
    camera_processing_consent: false,
    face: {
      consent_given: false,
      face_mode: "default_face"
    },
    aha_demo: {
      stage: "idle",
      lucy_preview_limit_sec: 10,
      google_generation_status: "idle",
      narrative: "Waiting for a recommendation before the realtime-to-high-quality try-on handoff starts."
    },
    lucy_session_status: "idle",
    loop_status: "idle",
    tool_calls: [],
    errors: []
  };
}
